table/builder.go

```diff
@@ -134,37 +124,64 @@ func (b *Builder) addHelper(key []byte, v y.ValueStruct) {
 		plen: uint16(len(key) - len(diffKey)),
 		klen: uint16(len(diffKey)),
 		vlen: uint16(v.EncodedSize()),
-		prev: b.prevOffset, // prevOffset is the location of the last key-value added.
 	}
-	b.prevOffset = uint32(b.buf.Len()) - b.baseOffset // Remember current offset for the next Add call.
+
+	// store current entries offset
+	b.blockEntryOffsets = append(b.blockEntryOffsets, uint32(b.buf.Len())-b.baseOffset)
 
 	// Layout: header, diffKey, value.
-	var hbuf [10]byte
+	var hbuf [6]byte
 	h.Encode(hbuf[:])
 	b.buf.Write(hbuf[:])
 	b.buf.Write(diffKey) // We only need to store the key difference.
 
 	v.EncodeTo(b.buf)
-	b.counter++ // Increment number of keys added for this current block.
 }
 
 func (b *Builder) finishBlock() {
-	// When we are at the end of the block and Valid=false, and the user wants to do a Prev,
-	// we need a dummy header to tell us the offset of the previous key-value pair.
-	b.addHelper([]byte{}, y.ValueStruct{})
+	// add entryOffsets at the end
+	offsetBuf := make([]byte, len(b.blockEntryOffsets)*4+4)
+	buf := offsetBuf
+
+	// put this in reverse order, while reading it will be picked in normal order
+	for i := len(b.blockEntryOffsets) - 1; i >= 0; i-- {
+		binary.BigEndian.PutUint32(buf[:4], b.blockEntryOffsets[i])
+		buf = buf[4:]
+	}
+	binary.BigEndian.PutUint32(buf[:4], uint32(len(b.blockEntryOffsets)))
+	b.buf.Write(offsetBuf)
+
+	// TODO: implement padding ??
+}
+
+func (b *Builder) shouldFinish(key []byte, value y.ValueStruct) bool {
```

Comment:
It would be better to call this function `shouldFinishBlock`. `shouldFinish` could be confused with `ReachedCapacity`.